dbms/src/Common/ThreadedWorker.h

```diff
@@ -0,0 +1,217 @@
+#include <mutex>
+
+namespace DB
+{
+
+/// A ThreadedWorker spawns n (n=concurrency) threads to process tasks.
+/// Each thread takes a task from the source queue, do some work, and put the result into the
+/// result queue.
+/// ThreadedWorker manages the state of its result queue, for example, when all tasks are finished
+/// or when there are errors, the result queue will be finished or cancelled.
+template <typename Src, typename Dest>
+class ThreadedWorker
+{
+public:
+    virtual ~ThreadedWorker()
+    {
+        wait();
+    }
+
+    void startInBackground() noexcept
+    {
+        std::call_once(start_flag, [this] {
+            LOG_DEBUG(log, "Starting {} workers, concurrency={}", getName(), concurrency);
+            active_workers = concurrency;
+
+            watch_start.restart();
+            for (size_t index = 0; index < concurrency; ++index)
+            {
+                thread_manager->schedule(true, getName(), [this, index] {
+                    total_wait_schedule_ms += watch_start.elapsedMilliseconds();
+
+                    workerLoop(index);
+                    handleWorkerFinished();
+                });
+            }
+        });
+    }
+
+    void wait() noexcept
+    {
+        std::call_once(wait_flag, [this] {
+            try
+            {
+                // thread_manager->wait can be only called once.
+                thread_manager->wait();
+            }
+            catch (...)
+            {
+                // This should not occur, as we should have caught all exceptions in workerLoop.
+                auto error = getCurrentExceptionMessage(false);
+                LOG_WARNING(log, "{} meet unexepcted error: {}", getName(), error);
+            }
+        });
+    }
+
+    ThreadedWorker(
+        std::shared_ptr<MPMCQueue<Src>> source_queue_,
+        std::shared_ptr<MPMCQueue<Dest>> result_queue_,
+        LoggerPtr log_,
+        size_t concurrency_)
+        : concurrency(concurrency_)
+        , source_queue(source_queue_)
+        , result_queue(result_queue_)
+        , log(log_)
+        , thread_manager(newThreadManager())
+    {
+        RUNTIME_CHECK(concurrency > 0);
+    }
+
+public:
+    const size_t concurrency;
+    const std::shared_ptr<MPMCQueue<Src>> source_queue;
+    const std::shared_ptr<MPMCQueue<Dest>> result_queue;
+
+protected:
+    const LoggerPtr log;
+    std::shared_ptr<ThreadManager> thread_manager;
+    std::atomic<Int64> active_workers = 0;
+
+    virtual String getName() const noexcept = 0;
+
+    virtual Dest doWork(const Src & task) = 0;
+
+private:
+    void handleWorkerFinished()
+    {
+        active_workers--;
+        if (active_workers == 0)
+        {
+            std::call_once(finish_flag, [this] {
+                LOG_DEBUG(
+                    log,
+                    "{} workers finished, total_processed_tasks={} concurrency={} elapsed={:.2f}s total_wait_schedule={:.2f}s total_wait_upstream={:.2f}s total_wait_downstream={:.2f}s",
```

Comment:
maybe `.3f` is better cause it corresponds to "milliseconds"
```suggestion
                    "{} workers finished, total_processed_tasks={} concurrency={} elapsed={:.3f}s total_wait_schedule={:.3f}s total_wait_upstream={:.3f}s total_wait_downstream={:.3f}s",
```